zilliqa/src/contracts/deposit.sol

```diff
@@ -138,123 +366,173 @@ contract Deposit {
+        futureCommittee.balances[uint(staker.keyIndex)] += msg.value;
     }
 
-    // temporary function to test liquid staking
-    function tempDecreaseDeposit(
-        bytes calldata blsPubKey,
-        uint256 amount
-    ) public {
-        Staker storage staker = _stakersMap[blsPubKey];
-        require(staker.keyIndex != 0, "unknown staker");
-        require(staker.rewardAddress == msg.sender, "invalid sender");
-        staker.balance -= amount;
+    function unstake(uint256 amount) public {
+        bytes storage stakerKey = _stakerKeys[msg.sender];
+        require(stakerKey.length != 0, "staker does not exist");
+        Staker storage staker = _stakersMap[stakerKey];
+
+        updateLatestComputedEpoch();
+
+        Committee storage futureCommittee = _committee[
+            (currentEpoch() + 2) % 3
+        ];
+
+        require(futureCommittee.stakers.length > 1, "too few stakers");
+
         require(
-            staker.balance == 0 || staker.balance >= _minimumStake,
-            "stake too low"
+            futureCommittee.balances[uint(staker.keyIndex)] >= amount,
+            "amount is greater than staked balance"
         );
-        totalStake -= amount;
-        (bool success, ) = msg.sender.call{value: amount}("");
-        require(success, "withdrawal failed");
-    }
-
-    function setStake(
-        bytes calldata blsPubKey,
-        bytes calldata peerId,
-        address rewardAddress,
-        uint256 amount
-    ) public {
-        require(msg.sender == address(0));
-        require(blsPubKey.length == 48);
-        require(peerId.length == 38);
 
-        if (amount < _minimumStake) {
-            revert("stake less than minimum stake");
+        if (futureCommittee.balances[uint(staker.keyIndex)] - amount == 0) {
+            // Remove the staker from the future committee, because their staked amount has gone to zero.
+            futureCommittee.totalStake -= amount;
+
+            // Delete this staker. We need to delete it in 3 places:
+            // 1. `futureCommittee.stakers`
+            // 2. `futureCommittee.balances`
+            // 3. `_stakersMap`
+
+            // First move the last staker into the position of the staker we want to delete. This needs to be done in
+            // `futureCommittee.stakers` and `futureCommittee.balances`. We also need to update the `keyIndex` of the
+            // last staker to point to its new index.
+            bytes storage lastStakerKey = futureCommittee.stakers[
+                futureCommittee.stakers.length - 1
+            ];
+            uint256 lastStakerBalance = futureCommittee.balances[
+                futureCommittee.balances.length - 1
+            ];
+            Staker storage lastStaker = _stakersMap[lastStakerKey];
+            futureCommittee.stakers[uint(staker.keyIndex)] = lastStakerKey;
+            futureCommittee.balances[uint(staker.keyIndex)] = lastStakerBalance;
+            lastStaker.keyIndex = staker.keyIndex;
```

Comment:
As you mention in the FIXME below were setting the `keyIndex` on `lastStaker` too early. Whenever we do so we also need to set `staker.keyIndex` to something else, perhaps -1. Not sure if its omitted here on purpose or forgotten